sound/hda/intel-nhlt.c

```diff
@@ -47,6 +47,33 @@ int intel_nhlt_get_dmic_geo(struct device *dev, struct nhlt_acpi_table *nhlt)
 			if (cfg->device_config.config_type != NHLT_CONFIG_TYPE_MIC_ARRAY) {
 				dev_warn(dev, "Generic config type, assuming single microphone\n");
 				dmic_geo = 1;
+
+				/* additional check based on format_configuration */
+				if (epnt->fmt_configs.fmt_count) {
+					u16 max_ch = 0;
+					u8 i;
+
+					dev_warn(dev, "Found %d format definitions\n",
+						 epnt->fmt_configs.fmt_count);
+
+					for (i = 0; i < epnt->fmt_configs.fmt_count; i++) {
+						struct wav_fmt_ext *fmt_ext;
+
+						fmt_ext = &epnt->fmt_configs.fmt_config[i].fmt_ext;
+
+						if (fmt_ext->fmt.channels > max_ch)
+							max_ch = fmt_ext->fmt.channels;
+					}
+					dev_warn(dev, "Max channels found %d\n", max_ch);
+					dmic_geo = max_ch;
```

Comment:
Using the format info is a bit iffy as it's not really describing the hw topology anymore, but rather at least all the NHLT spec examples show this being used to expose different DSP configurations (e.g. providing an option to get stereo stream for a 4mic array).

But if we go down this path, then looking up the max channel count makes sense. Would be strange for NHLT to not provide any configuration where all mics of the system are used. So seems good.

Reply:
yes, the recommendation was to lookup the maximum number of channels. I also think it's very odd but that's all we have.